### PR TITLE
Improve KVStore tests

### PR DESCRIPTION
**Pull request by CalvinNeo**
### What problem does this PR solve?

Issue Number: ref #7256

Problem Summary:

1. fix unstable snapshot tests.
2. improve read index tests.
3. introduce new util functions.
4. old snapshot test don't handle real data. fixed in new snapshot test. However, we will add asserts to show that.
5. move lock parsing codes into a seperated cpp file, will later investigate whether to use ffi to replace it.
6. rename some functions with extra copy. some of those calls should be reimpled to reduce copy.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Debug/MockRaftStoreProxy.cpp [hunk 1]
```diff
@@ -347,17 +347,18 @@ void MockRaftStoreProxy::testRunNormal(const std::atomic_bool & over)
...
+    while (!read_index_tasks.empty())
     {
-        auto & t = *tasks.front();
-        if (!region_id_to_drop.count(t.req.context().region_id()))
+        auto & t = *read_index_tasks.front();
+        auto region_id = t.req.context().region_id();
+        if (!region_id_to_drop.count(region_id))
         {
-            if (region_id_to_error.count(t.req.context().region_id()))
+            if (region_id_to_error.count(region_id))
```
Comment (Lloyd-Pottiger) [inline dbms/src/Debug/MockRaftStoreProxy.cpp:356]:
```suggestion
            if (region_id_to_error.contains(region_id))
```

File: dbms/src/Debug/MockRaftStoreProxy.cpp [hunk 2]
```diff
@@ -347,17 +347,18 @@ void MockRaftStoreProxy::testRunNormal(const std::atomic_bool & over)
...
 {
     auto _ = genLockGuard();
-    while (!tasks.empty())
+    while (!read_index_tasks.empty())
     {
-        auto & t = *tasks.front();
-        if (!region_id_to_drop.count(t.req.context().region_id()))
+        auto & t = *read_index_tasks.front();
+        auto region_id = t.req.context().region_id();
+        if (!region_id_to_drop.count(region_id))
```
Comment (Lloyd-Pottiger) [inline dbms/src/Debug/MockRaftStoreProxy.cpp:354]:
```suggestion
        if (!region_id_to_drop.contains(region_id))
```

File: dbms/src/Debug/MockRaftStoreProxy.cpp [hunk 3]
```diff
@@ -367,24 +368,46 @@ void MockRaftStoreProxy::unsafeInvokeForTest(std::function<void(MockRaftStorePro
...
     TMTContext & tmt,
     UInt64 region_id,
     std::optional<std::pair<std::string, std::string>> maybe_range)
+{
+    {
+        auto _ = genLockGuard();
+        if (!regions.empty())
+        {
+            throw Exception("Mock Proxy regions are not cleared");
+        }
```
Comment (Lloyd-Pottiger) [inline dbms/src/Debug/MockRaftStoreProxy.cpp:382]:
```suggestion
        RUNTIME_CHECK_MSG(regions.empty(), "Mock Proxy regions are not cleared");
```

File: dbms/src/Debug/MockRaftStoreProxy.cpp [hunk 4]
```diff
@@ -367,24 +368,46 @@ void MockRaftStoreProxy::unsafeInvokeForTest(std::function<void(MockRaftStorePro
...
+        auto task_lock = kvs.genTaskLock();
+        auto lock = kvs.genRegionWriteLock(task_lock);
+        if (!lock.regions.empty())
+        {
+            throw Exception("KVStore regions are not cleared");
+        }
+    }
+    auto start = RecordKVFormat::genKey(table_id, 0);
+    auto end = RecordKVFormat::genKey(table_id + 1, 0);
+    debugAddRegions(kvs, tmt, {region_id}, {maybe_range.value_or(std::make_pair(start.toString(), end.toString()))});
```
Comment (Lloyd-Pottiger) [inline dbms/src/Debug/MockRaftStoreProxy.cpp:392]:
```suggestion
    debugAddRegions(kvs, tmt, {region_id}, {maybe_range.value_or({start.toString(), end.toString()})});
```

File: dbms/src/Debug/MockRaftStoreProxy.cpp [hunk 5]
```diff
@@ -367,24 +368,46 @@ void MockRaftStoreProxy::unsafeInvokeForTest(std::function<void(MockRaftStorePro
...
+    std::vector<std::pair<std::string, std::string>> && ranges)
 {
     UNUSED(tmt);
+    int n = ranges.size();
     auto _ = genLockGuard();
-    regions.emplace(region_id, std::make_shared<MockProxyRegion>(region_id));
-
     auto task_lock = kvs.genTaskLock();
     auto lock = kvs.genRegionWriteLock(task_lock);
+    for (int i = 0; i < n; i++)
```
Comment (Lloyd-Pottiger) [inline dbms/src/Debug/MockRaftStoreProxy.cpp:406]:
```suggestion
    for (int i = 0; i < n; ++i)
```

File: dbms/src/Storages/Transaction/tests/kvstore_helper.h [hunk 6]
```diff
@@ -135,12 +138,12 @@ class RegionKVStoreTest : public ::testing::Test
 protected:
     static void testRaftSplit(KVStore & kvs, TMTContext & tmt);
     static void testRaftMerge(KVStore & kvs, TMTContext & tmt);
-    static void testRaftChangePeer(KVStore & kvs, TMTContext & tmt);
     static void testRaftMergeRollback(KVStore & kvs, TMTContext & tmt);
 
     static std::unique_ptr<PathPool> createCleanPathPool(const String & path)
     {
         // Drop files on disk
+        LOG_INFO(&Poco::Logger::get("Test"), "Clean path {} for bootstrap", path);
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/Transaction/tests/kvstore_helper.h:146]:
```suggestion
        LOG_INFO(Logger::get("Test"), "Clean path {} for bootstrap", path);
```

File: dbms/src/Debug/MockRaftStoreProxy.cpp (hunk 5 above)
Comment (CalvinNeo) [reply dbms/src/Debug/MockRaftStoreProxy.cpp:406]: Why?

File: dbms/src/Debug/MockRaftStoreProxy.cpp (hunk 4 above)
Comment (CalvinNeo) [reply dbms/src/Debug/MockRaftStoreProxy.cpp]: It can not compile, I revert this

File: dbms/src/Debug/MockRaftStoreProxy.cpp (hunk 5 above)
Comment (Lloyd-Pottiger) [reply dbms/src/Debug/MockRaftStoreProxy.cpp:403]: without compiler optimization, `++i` has better performance than `I++`

File: dbms/src/Debug/MockRaftStoreProxy.cpp [hunk 7]
```diff
@@ -497,6 +507,103 @@ std::tuple<uint64_t, uint64_t> MockRaftStoreProxy::compactLog(UInt64 region_id,
...
+std::tuple<uint64_t, uint64_t> MockRaftStoreProxy::compactLog(UInt64 region_id, UInt64 compact_index)
+{
+    auto region = getRegion(region_id);
+    assert(region != nullptr);
+    raft_cmdpb::AdminRequest request;
+    raft_cmdpb::AdminResponse response;
+    request.set_cmd_type(raft_cmdpb::AdminCmdType::CompactLog);
+    request.mutable_compact_log()->set_compact_index(compact_index);
+    // Find compact term, otherwise log must have been compacted.
+    if (region->commands.count(compact_index))
```
Comment (Lloyd-Pottiger) [inline dbms/src/Debug/MockRaftStoreProxy.cpp:519]:
```suggestion
    if (region->commands.contains(compact_index))
```

File: dbms/src/Debug/MockRaftStoreProxy.cpp [hunk 8]
```diff
@@ -497,6 +507,103 @@ std::tuple<uint64_t, uint64_t> MockRaftStoreProxy::compactLog(UInt64 region_id,
...
+    auto n = region_ids.size();
+    raft_cmdpb::AdminRequest request;
+    raft_cmdpb::AdminResponse response;
+    request.set_cmd_type(raft_cmdpb::AdminCmdType::BatchSplit);
+    metapb::RegionEpoch new_epoch;
+    new_epoch.set_version(old_epoch.version() + 1);
+    new_epoch.set_conf_ver(old_epoch.conf_ver());
+    {
+        raft_cmdpb::BatchSplitResponse * splits = response.mutable_splits();
+        for (size_t i = 0; i < n; i++)
```
Comment (Lloyd-Pottiger) [inline dbms/src/Debug/MockRaftStoreProxy.cpp:594]:
```suggestion
        for (size_t i = 0; i < n; ++i)
```

**Review by Lloyd-Pottiger — APPROVED** [review]
Others LGTM

**Review by hongyunyan — APPROVED** [review]
LGTM

**Pull request merged**